File: sdk/src/transaction.ts
import { TransactionData } from './types';

/**
 * HyperNova Chain transaction
 */
export class Transaction {
  /** Transaction type */
  public type?: string;
  /** Sender address */
  public from: string;
  /** Recipient address */
  public to: string;
  /** Amount */
  public amount: string;
  /** Transaction data */
  public data?: string;
  /** Transaction fee */
  public fee?: string;
  /** Nonce */
  public nonce?: number;
  /** Timestamp */
  public timestamp?: number;
  /** Signature */
  public signature?: string;
  /** Quantum-resistant signature */
  public quantumSignature?: string;

  /**
   * Create a new transaction
   * @param data Transaction data
   */
  constructor(data: TransactionData) {
    this.type = data.type;
    this.from = data.from;
    this.to = data.to;
    this.amount = data.amount;
    this.data = data.data;
    this.fee = data.fee;
    this.nonce = data.nonce;
    this.timestamp = data.timestamp || Math.floor(Date.now() / 1000);
    this.signature = data.signature;
    this.quantumSignature = data.quantumSignature;
  }

  /**
   * Convert to JSON
   * @returns Transaction as JSON
   */
  toJSON(): TransactionData {
    return {
      type: this.type,
      from: this.from,
      to: this.to,
      amount: this.amount,
      data: this.data,
      fee: this.fee,
      nonce: this.nonce,
      timestamp: this.timestamp,
      signature: this.signature,
      quantumSignature: this.quantumSignature,
    };
  }

  /**
   * Create a token transfer transaction
   * @param from Sender address
   * @param to Recipient address
   * @param amount Amount to send
   * @returns Transaction
   */
  static createTransfer(from: string, to: string, amount: string): Transaction {
    return new Transaction({
      type: 'transfer',
      from,
      to,
      amount,
    });
  }

  /**
   * Create a contract deployment transaction
   * @param from Deployer address
   * @param bytecode Contract bytecode
   * @param amount Amount to send
   * @returns Transaction
   */
  static createContractDeployment(from: string, bytecode: string, amount: string = '0'): Transaction {
    return new Transaction({
      type: 'contract_deploy',
      from,
      to: '',
      amount,
      data: bytecode,
    });
  }

  /**
   * Create a contract call transaction
   * @param from Caller address
   * @param to Contract address
   * @param data Call data
   * @param amount Amount to send
   * @returns Transaction
   */
  static createContractCall(from: string, to: string, data: string, amount: string = '0'): Transaction {
    return new Transaction({
      type: 'contract_call',
      from,
      to,
      amount,
      data,
    });
  }
}

File: core/src/transaction.rs
use serde::{Serialize, Deserialize};
use sha2::{Sha256, Digest};
use std::time::{SystemTime, UNIX_EPOCH};

use crate::crypto::Signature;

/// Transaction types supported by HyperNova Chain
#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum TransactionType {
    /// Transfer of tokens
    Transfer,
    /// Smart contract deployment
    ContractDeploy,
    /// Smart contract call
    ContractCall,
    /// Validator registration
    ValidatorRegistration,
    /// AI model submission (for PoAI)
    AIModelSubmission,
    /// Governance proposal
    GovernanceProposal,
    /// Governance vote
    GovernanceVote,
}

/// Transaction structure for HyperNova Chain
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Transaction {
    /// Transaction type
    pub tx_type: TransactionType,
    /// Sender's address
    pub from: String,
    /// Recipient's address (if applicable)
    pub to: Option<String>,
    /// Amount of tokens (if applicable)
    pub amount: Option<u64>,
    /// Transaction data (e.g., contract bytecode, call data)
    pub data: Option<Vec<u8>>,
    /// Transaction fee
    pub fee: u64,
    /// Nonce to prevent replay attacks
    pub nonce: u64,
    /// Timestamp when the transaction was created
    pub timestamp: u64,
    /// Sender's signature
    pub signature: Option<Signature>,
    /// Quantum-resistant signature (if enabled)
    pub quantum_signature: Option<Vec<u8>>,
}

impl Transaction {
    /// Create a new transaction
    pub fn new(
        tx_type: TransactionType,
        from: String,
        to: Option<String>,
        amount: Option<u64>,
        data: Option<Vec<u8>>,
        fee: u64,
        nonce: u64,
    ) -> Self {
        let timestamp = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .unwrap()
            .as_secs();
            
        Transaction {
            tx_type,
            from,
            to,
            amount,
            data,
            fee,
            nonce,
            timestamp,
            signature: None,
            quantum_signature: None,
        }
    }
    
    /// Calculate the hash of this transaction
    pub fn hash(&self) -> String {
        // Create a copy without signatures for hashing
        let mut tx_copy = self.clone();
        tx_copy.signature = None;
        tx_copy.quantum_signature = None;
        
        let tx_json = serde_json::to_string(&tx_copy).unwrap();
        let mut hasher = Sha256::new();
        hasher.update(tx_json.as_bytes());
        format!("{:x}", hasher.finalize())
    }
    
    /// Sign the transaction with sender's signature
    pub fn sign(&mut self, signature: Signature) {
        self.signature = Some(signature);
    }
    
    /// Add quantum-resistant signature
    pub fn add_quantum_signature(&mut self, signature: Vec<u8>) {
        self.quantum_signature = Some(signature);
    }
    
    /// Verify the transaction signature
    pub fn verify_signature(&self) -> bool {
        // In a real implementation, this would verify the signature
        // against the sender's public key
        self.signature.is_some()
    }
}